api/api.go

```diff
@@ -1,187 +1,136 @@
+		return nil, nil, driverOpts, errors.GetDriver.Combine(err)
 	}
 
-	uc := &config.UseMount{
-		Volume:   strings.Join([]string{policy, volume}, "/"),
-		Reason:   lock.ReasonCreate,
-		Hostname: hostname,
-	}
-
-	snapUC := &config.UseSnapshot{
-		Volume: strings.Join([]string{policy, volume}, "/"),
-		Reason: lock.ReasonCreate,
-	}
-
-	global := *a.Global
-
-	err = lock.NewDriver(a.Client).ExecuteWithMultiUseLock([]config.UseLocker{uc, snapUC}, global.Timeout, func(ld *lock.Driver, ucs []config.UseLocker) error {
-		volConfig, err := a.Client.CreateVolume(config.Request{Policy: policy, Volume: volume, Options: req.Opts})
-		if err != nil {
-			return err
-		}
-
-		log.Debugf("Volume Create: %#v", *volConfig)
-
-		do, err := control.CreateVolume(policyObj, volConfig, global.Timeout)
-		if err == errors.NoActionTaken {
-			goto publish
-		}
-
-		if err != nil {
-			return errors.CreateVolume.Combine(err)
-		}
-
-		if err := control.FormatVolume(volConfig, do); err != nil {
-			if err := control.RemoveVolume(volConfig, global.Timeout); err != nil {
-				log.Errorf("Error during cleanup of failed format: %v", err)
-			}
-			return errors.FormatVolume.Combine(err)
-		}
-
-	publish:
-		if err := a.Client.PublishVolume(volConfig); err != nil && err != errors.Exists {
-			// FIXME this shouldn't leak down to the client.
-			if _, ok := err.(*errored.Error); !ok {
-				return errors.PublishVolume.Combine(err)
-			}
-			return err
-		}
-
-		content, err = json.Marshal(volConfig)
-		if err != nil {
-			return errors.MarshalPolicy.Combine(err)
-		}
-
-		w.Write(content)
-		return nil
-	})
-
-	if err != nil && err != errors.Exists {
-		a.HTTPError(w, errors.CreateVolume.Combine(err))
-		return
+	driverOpts, err = volConfig.ToDriverOptions((*a.Global).Timeout)
+	if err != nil {
+		return nil, nil, driverOpts, errors.UnmarshalRequest.Combine(err)
 	}
-}
 
-// HTTPError is a generic HTTP error function that works across the plugin and
-// REST interfaces. It is intended to be used by handlers that exist in this
-// package.
-func (a *API) HTTPError(w http.ResponseWriter, err error) {
-	if a.DockerPlugin {
-		DockerHTTPError(w, err)
-	} else {
-		RESTHTTPError(w, err)
-	}
+	return driver, volConfig, driverOpts, nil
 }
 
-// DockerHTTPError returns a 200 status to docker with an error struct. It returns
-// 500 if marshalling failed.
-func DockerHTTPError(w http.ResponseWriter, err error) {
-	content, errc := json.Marshal(VolumeCreateResponse{Mountpoint: "", Err: err.Error()})
-	if errc != nil {
-		http.Error(w, errc.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	log.Errorf("Returning HTTP error handling plugin negotiation: %s", err.Error())
-	http.Error(w, string(content), http.StatusOK)
+// AddStopChan adds a stop channel for the purposes of controlling mount ttl refresh goroutines
+func (a *API) AddStopChan(name string, stopChan chan struct{}) {
+	a.lockStopChanMutex.Lock()
+	a.lockStopChans[name] = stopChan
+	a.lockStopChanMutex.Unlock()
```

Comment:
another way of putting this: in the extremely unlikely event the panic does happen and it doesn't make it to main (which will abort the program), I'd rather have that mutex hang everything instead of it taking further action. Something is very f'd up if it does.